### PR TITLE
CLN: type annotations in groupby.grouper, groupby.ops

### PR DESCRIPTION
**Pull request by jbrockmendel**
@simonjayhawkins mypy is still giving a couple of complaints I could use your help sorting out:

```
pandas/core/groupby/ops.py:791: error: Signature of "groupings" incompatible with supertype "BaseGrouper"
pandas/core/groupby/ops.py:872: error: Argument 1 of "_chop" is incompatible with supertype "DataSplitter"; supertype defines the argument type as "NDFrame"
pandas/core/groupby/ops.py:884: error: Argument 1 of "_chop" is incompatible with supertype "DataSplitter"; supertype defines the argument type as "NDFrame"
```

For the groupings complaint, AFAICT the attribute has the same annotation, but in the subclass its a property instead of defined in `__init__`.  For the other two, I annotated an argument with `NDFrame` in the base class and overrode with `Series` and `DataFrame` in the subclasses.  What is the preferred idiom for this pattern?

### CODE REVIEW

File: pandas/core/groupby/ops.py [hunk 1]
```diff
@@ -89,7 +89,7 @@ def __init__(
 
         self._filter_empty_groups = self.compressed = len(groupings) != 1
         self.axis = axis
-        self.groupings = groupings  # type: Sequence[grouper.Grouping]
+        self.groupings = list(groupings)  # type: List[grouper.Grouping]
```
Comment (WillAyd) [inline pandas/core/groupby/ops.py]: Was this necessary? I think Sequence should have been allowable

File: pandas/core/groupby/grouper.py [hunk 2]
```diff
@@ -144,17 +144,18 @@ def _get_grouper(self, obj, validate=True):
         )
         return self.binner, self.grouper, self.obj
 
-    def _set_grouper(self, obj, sort=False):
+    def _set_grouper(self, obj: NDFrame, sort: bool = False):
```
Comment (WillAyd) [inline pandas/core/groupby/grouper.py]: Return value can be annotated as None here right?

File: pandas/core/groupby/ops.py [hunk 3]
```diff
@@ -157,7 +157,7 @@ def apply(self, f, data, axis: int = 0):
 
         elif (
             com.get_callable_name(f) not in base.plotting_methods
-            and hasattr(splitter, "fast_apply")
+            and splitter.fast_apply is not None
```
Comment (WillAyd) [inline pandas/core/groupby/ops.py]: I can see why mypy complains about this but the new way of doing this might be fragile as well. Would making this an abstract property in the base class be difficult?

File: pandas/core/groupby/ops.py [hunk 4]
```diff
@@ -825,7 +827,9 @@ def _is_indexed_like(obj, axes) -> bool:
 
 
 class DataSplitter:
-    def __init__(self, data, labels, ngroups, axis: int = 0):
+    fast_apply = None
+
+    def __init__(self, data: NDFrame, labels, ngroups: int, axis: int = 0):
```
Comment (WillAyd) [inline pandas/core/groupby/ops.py]: I *think* this is a generic implementation, so should use `FrameOrSeries` instead of `NDFrame` (may help some of the other issues from your OP)

**Review by WillAyd — CHANGES_REQUESTED** [review]
looks good some comments

File: pandas/core/groupby/ops.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/groupby/ops.py]: This was an attempt to troubleshoot a (still-failing, mentioned in OP) complaint about a mismatch in the `groupings` annotations.

File: pandas/core/groupby/grouper.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/groupby/grouper.py]: Will double-check and add if so

File: pandas/core/groupby/ops.py (hunk 3 above)
Comment (jbrockmendel) [reply pandas/core/groupby/ops.py]: I dont think mypy is complaining about this, but ill change this check to be clearer on that front

File: pandas/core/groupby/ops.py (hunk 4 above)
Comment (jbrockmendel) [reply pandas/core/groupby/ops.py]: I just tried changing all the `NDFrame` annotations to `FrameOrSeries` and now I'm getting additional complaints:

```
pandas/core/groupby/ops.py:869: error: Return type "Series" of "_chop" incompatible with return type "FrameOrSeries" in supertype "DataSplitter"
pandas/core/groupby/ops.py:881: error: Return type "DataFrame" of "_chop" incompatible with return type "FrameOrSeries" in supertype "DataSplitter"
```

Was this comment supposed to just apply to the `__init__` on 832?  Regardless, WTF?

Comment (WillAyd) [reply pandas/core/groupby/ops.py]: I was wrong before; I guess as a rule of thumb if it gets subclassed OK to type as NDFrame in superclass and DataFrame / Series in appropriate subclass. However, if it doesn't get subclassed it's probably a generic function and should utilize FrameOrSeries

File: pandas/core/groupby/grouper.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/groupby/grouper.py]: no, it returns self.grouper

File: pandas/core/groupby/ops.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/groupby/ops.py]: @simonjayhawkins any thoughts on why mypy is still complaining about groupings?

Comment (simonjayhawkins) [reply pandas/core/groupby/ops.py]: I think you're right about the property and the attribute. 

```
diff --git a/pandas/core/groupby/ops.py b/pandas/core/groupby/ops.py
index cbe012012..46a427434 100644
--- a/pandas/core/groupby/ops.py
+++ b/pandas/core/groupby/ops.py
@@ -90,12 +90,16 @@ class BaseGrouper:
 
         self._filter_empty_groups = self.compressed = len(groupings) != 1
         self.axis = axis
-        self.groupings = list(groupings)  # type: List[grouper.Grouping]
+        self._groupings = list(groupings)  # type: List[grouper.Grouping]
         self.sort = sort
         self.group_keys = group_keys
         self.mutated = mutated
         self.indexer = indexer
 
+    @property
+    def groupings(self):
+        return self._groupings
+
     @property
     def shape(self):
         return tuple(ping.ngroups for ping in self.groupings)
```

silences mypy if you're happy with this approach.

Comment (jbrockmendel) [reply pandas/core/groupby/ops.py]: seems to work, updated

**Comment by jbrockmendel** [comment]
Any issues remaining here?

File: pandas/core/groupby/grouper.py [hunk 5]
```diff
@@ -143,17 +143,18 @@ def _get_grouper(self, obj, validate=True):
...
         for this particular specification
 
         Parameters
         ----------
-        obj : the subject object
+        obj : Series or DataFrame
         sort : bool, default False
             whether the resulting grouper should be sorted
         """
+        assert obj is not None
```
Comment (simonjayhawkins) [inline pandas/core/groupby/grouper.py]: why is this needed?

File: pandas/core/groupby/ops.py [hunk 6]
```diff
@@ -142,13 +147,13 @@ def _get_group_keys(self):
...
 
-    def apply(self, f, data, axis: int = 0):
+    def apply(self, f, data: FrameOrSeries, axis: int = 0):
         mutated = self.mutated
         splitter = self._get_splitter(data, axis=axis)
         group_keys = self._get_group_keys()
         result_values = None
 
-        sdata = splitter._get_sorted_data()
+        sdata = splitter._get_sorted_data()  # type: FrameOrSeries
```
Comment (simonjayhawkins) [inline pandas/core/groupby/ops.py]: why is this needed?

shouldn't need to add a type annotation here. maybe the return type of `_get_sorted_data` needs to be added.

File: pandas/core/groupby/grouper.py (hunk 5 above)
Comment (jbrockmendel) [reply pandas/core/groupby/grouper.py]: This is for my own benefit in trying to reason about this code.

File: pandas/core/groupby/ops.py (hunk 6 above)
Comment (jbrockmendel) [reply pandas/core/groupby/ops.py]: _get_sorted_data return type is annotated, but mypy complains without this

File: pandas/core/groupby/ops.py [hunk 7]
```diff
@@ -861,15 +868,15 @@ def __iter__(self):
         for i, (start, end) in enumerate(zip(starts, ends)):
             yield i, self._chop(sdata, slice(start, end))
 
-    def _get_sorted_data(self):
+    def _get_sorted_data(self) -> FrameOrSeries:
         return self.data.take(self.sort_idx, axis=self.axis)
 
-    def _chop(self, sdata, slice_obj: slice):
+    def _chop(self, sdata, slice_obj: slice) -> NDFrame:
```
Comment (simonjayhawkins) [inline pandas/core/groupby/ops.py]: why is NDFrame used? is _chop not generic? should DataSplitter be a generic class?

Comment (jbrockmendel) [reply pandas/core/groupby/ops.py]: I dont understand the question.  Is "generic class" meaningfully different from "base class"?  NDFrame is used because one subclass returns Series and the other returns DataFrame

File: pandas/core/groupby/ops.py [hunk 8]
```diff
@@ -881,14 +888,14 @@ def fast_apply(self, f, names):
         sdata = self._get_sorted_data()
         return libreduction.apply_frame_axis0(sdata, f, names, starts, ends)
 
-    def _chop(self, sdata, slice_obj: slice):
+    def _chop(self, sdata: DataFrame, slice_obj: slice) -> DataFrame:
```
Comment (simonjayhawkins) [inline pandas/core/groupby/ops.py]: should this be Series?

**Comment by simonjayhawkins** [comment]
> Any issues remaining here?

I think should use py3.6 syntax for variable annotations going forward and would also prefer to not to see changes to .format but use f-strings instead.

File: pandas/core/groupby/ops.py (hunk 7 above)
Comment (simonjayhawkins) [reply pandas/core/groupby/ops.py]: `DataSplitter.__init__` accepts `FrameOrSeries`. do we need to persist this type thoughout the class. i.e. make DataSplitter a generic class. see https://mypy.readthedocs.io/en/latest/generics.html#defining-generic-classes

File: pandas/core/groupby/ops.py (hunk 8 above)
Comment (jbrockmendel) [reply pandas/core/groupby/ops.py]: im pretty sure this is correct as-is

**Comment by jbrockmendel** [comment]
> I think should use py3.6 syntax for variable annotations going forward and would also prefer to not to see changes to .format but use f-strings instead.

3.6 annotations ill try to get in the habit of.  f-strings are going to take some getting used to

**Comment by simonjayhawkins** [comment]
> f-strings are going to take some getting used to

just thinking it'll reduce churn as they are likely to be updated anyway. so why makes changes twice?

**Comment by jbrockmendel** [comment]
> just thinking it'll reduce churn as they are likely to be updated anyway. so why makes changes twice?

No reason at all.  That doesn't change the whole "I'm an old man and change will take some getting used to" thing.

File: pandas/core/groupby/ops.py (hunk 8 above)
Comment (simonjayhawkins) [reply pandas/core/groupby/ops.py]: ahh ok, had the class FrameSplitter hidden and just saw the class SeriesSplitter above. I guess sdata doesn't mean series data then.

File: pandas/core/groupby/ops.py (hunk 7 above)
Comment (simonjayhawkins) [reply pandas/core/groupby/ops.py]: so looking at the definition of _chop in the derived classes, i'm guessing this abstractmethod should be typed as 
```
 def _chop(self, sdata: FrameOrSeries, slice_obj: slice) -> FrameOrSeries:
```

Comment (jbrockmendel) [reply pandas/core/groupby/ops.py]: Using FrameOrSeries here produces complaints:

```
pandas/core/groupby/ops.py:879: error: Argument 1 of "_chop" is incompatible with supertype "DataSplitter"; supertype defines the argument type as "FrameOrSeries"
pandas/core/groupby/ops.py:879: error: Return type "Series" of "_chop" incompatible with return type "FrameOrSeries" in supertype "DataSplitter"
pandas/core/groupby/ops.py:891: error: Argument 1 of "_chop" is incompatible with supertype "DataSplitter"; supertype defines the argument type as "FrameOrSeries"
pandas/core/groupby/ops.py:891: error: Return type "DataFrame" of "_chop" incompatible with return type "FrameOrSeries" in supertype "DataSplitter"
```

I'm getting close to saying "screw it" when dealing with this type of error.

Comment (simonjayhawkins) [reply pandas/core/groupby/ops.py]: fair enough. probably best to remove the type annotation here then for now. Having one method return FrameOrSeries and another NDFrame seems inconsistent.

Comment (WillAyd) [reply pandas/core/groupby/ops.py]: I think `NDFrame` is correct here; this function isn't generic in the base class rather the subclasses override with the appropriate object

Comment (simonjayhawkins) [reply pandas/core/groupby/ops.py]: mypy won't be looking at the derived classes when it performs type checking. it'll be looking at the type hints on the base class when it checks other methods in the base class.

the abstractmethod should be generic since that is how the derived classes are typed Series -> Series and DataFrame -> DataFrame.

Comment (WillAyd) [reply pandas/core/groupby/ops.py]: I think we are mixing a few different paradigms here. The subclasses should probably be annotated with the type respective to the class, rather than using the TypeVar, i.e. you would never parametrize a `SeriesSplitter` with a DataFrame - it exclusively deals with Series objects

Comment (simonjayhawkins) [reply pandas/core/groupby/ops.py]: > you would never parametrize a `SeriesSplitter` with a DataFrame - it exclusively deals with Series objects

correct. but if a method of the base class is not overridden then the Series type in the derived class will become an NDFrame type after calling that method in the base class.

Comment (jbrockmendel) [reply pandas/core/groupby/ops.py]: > I think we are mixing a few different paradigms here.

There are some annotations in this PR that make it easier to reason about this code while reading it.  The annotations in this sub-thread are not among them, so I do not particularly care about them.  Let's focus for now on a minimal change needed to get this merged, as there are more bugfix PRs waiting in the wings.

**Comment by jbrockmendel** [comment]
rebased+green

File: pandas/core/groupby/ops.py (hunk 6 above)
Comment (jreback) [reply pandas/core/groupby/ops.py]: can update to py3.6 syntax in a followon

**Comment by jreback** [comment]
lgtm (ex one comment), @WillAyd

**Review by WillAyd — APPROVED** [review]

**Comment by WillAyd** [comment]
Thanks @jbrockmendel

File: pandas/core/groupby/ops.py (hunk 7 above)
Comment (simonjayhawkins) [reply pandas/core/groupby/ops.py]: https://github.com/pandas-dev/pandas/pull/28339/commits/e6c5f5a31e5a5912bd572e08f03756cbc075752c fixes this.

File: pandas/core/groupby/ops.py (hunk 6 above)
Comment (simonjayhawkins) [reply pandas/core/groupby/ops.py]: no longer needed after https://github.com/pandas-dev/pandas/pull/28339/commits/e6c5f5a31e5a5912bd572e08f03756cbc075752c
